bin/write_viral_gff.py

```diff
@@ -21,136 +20,213 @@ def get_ena_contig_mapping(ena_contig_file):
-        metadata.write(headers)
-        for contig in virify_contigs:
-            #   change contig name if ena mapping required
-            if ena_mapping:
-                contig_name = f"{ena_mapping[contig]}-{contig}"
+                    Contig=Contig.replace('|','')
+
+                lineage=[]
+                for element in line_l:
+                    if len(element)>0:
+                        try:
+                            float(element)
+                        except:
+                            lineage.append(element)
+                if len(lineage)>0:
+                    taxo_string='%3B'.join(lineage)
+                else:
+                    taxo_string='unclassified'
+
+                taxonomy_dict[Contig]=taxo_string
+
+    # Saving the original contig length
+    for record in SeqIO.parse(assembly_file, "fasta"):
+        contig_id=str(record.id)
+        seq_len=len(str(record.seq))
+        contigs_len_dict[contig_id]=str(seq_len)
+
+    contig_protID=annot_dicts[0]
+    prot_data=annot_dicts[1]
+    contig_data=annot_dicts[2]
+    # Writing the gff file
+    with open(output_filename, "w") as gff:
+        gff.write("##gff-version 3\n")
+        contig_protID=annot_dicts[0]
+        prot_data=annot_dicts[1]
+        contig_data=annot_dicts[2]
+
+        # Writing the gff header
+        printed_contig=[]
+        for element in contig_protID.keys():
+            if 'prophage' in element:
+                clean_id=element.split('|')[0]
+            else:
+                clean_id=element
+
+            if clean_id not in printed_contig:
+                printed_contig.append(clean_id)
+                element_len=contigs_len_dict[clean_id]
+                gff.write('##sequence-region '+clean_id+' 1 '+element_len+'\n')
+
+        # Writing the mobile genetic elements coordinates and attributes
+        phage_counter=0
+        phage_ids={}
+        for element in contig_protID.keys():
+            phage_counter+=1
+            element_id='phage_'+str(phage_counter)
+            phage_ids[element]=element_id
+            if 'prophage' in element:
+                seq_type='prophage'
+                seqid=element.split('|')[0]
+                start=element.split('|')[1].split('-')[1].split(':')[0]
+                if start=='0':
+                    start='1'
+                region_end=element.split('|')[1].split('-')[1].split(':')[1]
+                if int(region_end) > int(contigs_len_dict[seqid]):
+                    region_end=str(contigs_len_dict[seqid])
+            else:
+                seq_type='viral_sequence'
+                seqid=element
+                start='1'
+                region_end=contigs_len_dict[element]
+            
+            if element in taxonomy_dict.keys():
+                attributes='ID='+element_id+';gbkey=mobile_element;mobile_element_type='+contig_data[element]+';'+checkv_dict[element]+';taxonomy='+taxonomy_dict[element]
```

Comment:
Improved readability:
```suggestion
        attributes = ";".join([
            f"ID={element_id}",
            f"gbkey={mobile_element}",
            f"mobile_element_type={contig_data[element]}",
            f"checkv_dict[element]'",
            f"taxonomy={taxonomy_dict[element]}"
        ])
```